### PR TITLE
Add adminListServices endpoint

### PR DESCRIPTION
**Pull request by MongoCaleb**
### Jira

https://jira.mongodb.org/browse/DOCSP-33421

### Staged Changes

https://preview-mongodbmongocaleb.gatsbyjs.io/atlas-app-services/yaml_test/admin/api/v3/#tag/services/operation/adminListServices

## Note
The scope of this ticket expanded to including a new endpoint and updating the response objects for several endpoints.

### CODE REVIEW

File: source/openapi-admin-v3.yaml [hunk 1]
```diff
@@ -847,48 +847,70 @@ paths:
               schema:
                 items:
                   anyOf:
-                    - $ref: "#/components/schemas/AtlasCluster"
+                    - $ref: "#/components/schemas/ServiceResponse"
```
Comment (nlarew) [inline source/openapi-admin-v3.yaml:850]: I think we should remove the `anyOf` here and only use `ServiceResponse`

Screenshot 2023-10-12 at 11 39 38 AM

File: source/openapi-admin-v3.yaml [hunk 2]
```diff
@@ -5525,7 +5547,71 @@ components:
...
+        name: string;
+        type: string;
+        version: number;
+        last_modified: number;
+    ReadPreferencesTagSet:
+      type: string
+      enum:
+        - server1
+        - server2
+    CreateServiceRequest:
```
Comment (nlarew) [inline source/openapi-admin-v3.yaml:5614]: This seems to be specifically for data sources - can we use a less generic name? Also maybe an opportunity to reuse the `AtlasCluster` schema?

File: source/openapi-admin-v3.yaml [hunk 3]
```diff
@@ -5539,35 +5625,49 @@ components:
           enum:
             - mongodb-atlas
         config:
+          config: "null"
```
Comment (nlarew) [inline source/openapi-admin-v3.yaml:5628]: Typo? I'm not familiar with the `config` field and not sure what "null" would mean here

File: source/openapi-admin-v3.yaml [hunk 4]
```diff
@@ -5600,7 +5700,7 @@ components:
         - type
         - config
       oneOf:
-        - $ref: "#/components/schemas/AtlasCluster"
+        - $ref: "#/components/schemas/AppConstructor"
```
Comment (nlarew) [inline source/openapi-admin-v3.yaml:5703]: This one should still be `AtlasCluster`.

`AppConstructor` reference `DataSource` so this is circular as-is.

File: source/openapi-admin-v3.yaml [hunk 5]
```diff
@@ -5525,7 +5547,71 @@ components:
         error_code:
           type: string
           description: The error type.
-    AtlasCluster:
+    AtlasConfigResponse:
```
Comment (nlarew) [inline source/openapi-admin-v3.yaml:5550]: Can we still use `AtlasCluster` here? It's a bit more descriptive (this doesn't represent Atlas Federated Instances for example).

File: source/openapi-admin-v3.yaml [hunk 6]
```diff
@@ -5525,7 +5547,71 @@ components:
...
+          properties:
+            clusterId:
+              type: number
+              example: 05e821b3975d271289f372e3a
+              description: The service ID.
+            clusterName:
+              type: string
+              example: Cluster0
+              description: |-
+                The service name used to refer to the data source within
```
Comment (nlarew) [inline source/openapi-admin-v3.yaml:5564]: Cluster Name !== Service Name (though eventually they say they will be). This is the actual name of the cluster in Atlas, not e.g. "mongodb-atlas".

File: source/openapi-admin-v3.yaml [hunk 7]
```diff
@@ -5525,7 +5547,71 @@ components:
...
+                this Atlas app. The name may be at most 64 characters
+                long and must only contain ASCII letters, numbers,
+                underscores, and hyphens.
+            clusterType:
+              type: string
+              example: atlas
+              description: The type of service.
```
Comment (nlarew) [inline source/openapi-admin-v3.yaml:5571]:
```suggestion
            clusterType:
              type: string
              enum:
                - atlas
                - serverless
              example: atlas
              description: The type of service.
```

There's only two options here so we can use an enum

**Review by nlarew — CHANGES_REQUESTED** [review]

File: source/openapi-admin-v3.yaml (hunk 3 above)
Comment (MongoCaleb) [reply source/openapi-admin-v3.yaml:5628]: Forgot to remove that -- it was a placeholder until I figured out the details. :( Good catch!

File: source/openapi-admin-v3.yaml (hunk 1 above)
Comment (MongoCaleb) [reply source/openapi-admin-v3.yaml:850]: I'd love to know why.

File: source/openapi-admin-v3.yaml [hunk 8]
```diff
@@ -850,45 +869,46 @@ paths:
                     - $ref: "#/components/schemas/AtlasCluster"
                     - $ref: "#/components/schemas/AtlasFederatedInstance"
                     - $ref: "#/components/schemas/ThirdPartyService"
-    post:
+    patch:
```
Comment (nlarew) [inline source/openapi-admin-v3.yaml:872]: We lost the real `adminCreateService` endpoint somewhere here - there should be an endpoint for `POST /groups/{groupId}/apps/{appId}/services/config`.

Let's re-add that endpoint and then update this one to be `adminUpdateServiceConfig` or similar with an appropriate `summary`.

Comment (MongoCaleb) [reply source/openapi-admin-v3.yaml:872]: gah! Moving things around so much has led to quite the mess.

File: source/openapi-admin-v3.yaml [hunk 9]
```diff
@@ -882,13 +884,70 @@ paths:
...
+          description: Successfully created.
+          content:
+            application/json:
+              schema:
+                oneOf:
+                  - $ref: "#/components/schemas/ServiceResponse"
+                  - $ref: "#/components/schemas/AtlasFederatedInstance"
+                  - $ref: "#/components/schemas/ThirdPartyService"
+        "409":
+          description: There is already a service with the given `name`.
```
Comment (nlarew) [inline source/openapi-admin-v3.yaml:939]:
```suggestion
        "204":
          description: Successfully updated.
```

Looks like this still has the createService responses. Just checked and the update config endpoint returns 204 No Content

Screenshot 2023-10-13 at 5 40 47 PM

File: source/openapi-admin-v3.yaml [hunk 10]
```diff
@@ -898,7 +957,7 @@ paths:
             application/json:
               schema:
                 oneOf:
-                  - $ref: "#/components/schemas/AtlasCluster"
+                  - $ref: "#/components/schemas/AppConstructor"
```
Comment (nlarew) [inline source/openapi-admin-v3.yaml:960]:
```suggestion
                  - $ref: "#/components/schemas/AtlasCluster"
```

File: source/openapi-admin-v3.yaml (hunk 1 above)
Comment (nlarew) [reply source/openapi-admin-v3.yaml]: > I'd love to know why.

Per that screenshot, the endpoint only returns `ServiceResponse` objects

**Review by nlarew — APPROVED** [review]
LGTM

**Pull request merged**